File: packages/icp-python-client/icp_client/webhook.py
"""ICPIP-0005 receiver helpers — `verify_webhook` is the Stripe-style
one-call validator for inbound webhook events.

Hand it the raw HTTP body string, the request headers (case-insensitive),
the HTTP method/path, and the merchant's raw 32-byte Ed25519 pubkey from
`.well-known/icp`. Get back the parsed `EventEnvelope` dict, OR an
`ICPError` with a `channel.*` code that maps directly to HTTP status.

Performs every check ICPIP-0005 §6 mandates:
  1. HTTP timestamp within ±tolerance (default 300s) → `channel.replay` on miss.
  2. HTTP-layer `X-ICP-Signature: ed25519=<hex>` verifies against
     `<timestamp>.<method>.<path>.<body>`.
  3. Body parses as `{envelope, signature}`.
  4. Envelope signature verifies against the merchant pubkey over the
     envelope's canonical JSON bytes.

This mirrors the JavaScript SDK's `verifyWebhook` function byte-for-byte.
"""

from __future__ import annotations

import json
import re
import time
from typing import Any, Mapping

from .client import ICPError
from .codec import canonical_json, verify_ed25519


_HTTP_SIG_PATTERN = re.compile(r"^ed25519=([0-9a-fA-F]+)$")


def _normalized_header_lookup(headers: Mapping[str, str] | None, name: str) -> str | None:
    """Look up `name` (case-insensitive) in a header mapping.

    Accepts dict, multidict, requests.structures.CaseInsensitiveDict, or
    any object that exposes either case-insensitive `__getitem__` or
    standard `dict.get` semantics.
    """
    if not headers:
        return None
    # 1. Try exact key.
    if hasattr(headers, "get"):
        v = headers.get(name)
        if v is not None:
            return v
        v = headers.get(name.lower())
        if v is not None:
            return v
        v = headers.get(name.upper())
        if v is not None:
            return v
    # 2. Fall back to scan for case-insensitive match.
    target = name.lower()
    try:
        items = headers.items()  # type: ignore[union-attr]
    except AttributeError:
        return None
    for k, v in items:
        if k.lower() == target:
            return v
    return None


def verify_webhook(
    *,
    body: str,
    headers: Mapping[str, str],
    method: str,
    path: str,
    merchant_pubkey_raw: bytes,
    tolerance_seconds: int = 300,
    now_seconds: int | None = None,
) -> dict[str, Any]:
    """Verify an inbound webhook and return its parsed `EventEnvelope`.

    Raises `ICPError` (a typed exception with a `channel.*` code) on any
    verification failure.

    :param body: Raw HTTP body string. Do NOT pre-parse — JSON re-encoding
                 would break the HTTP-layer signature.
    :param headers: HTTP request headers (case-insensitive lookup).
    :param method: HTTP method (e.g. "POST").
    :param path: HTTP path (including query string if the original request
                 had one).
    :param merchant_pubkey_raw: Raw 32-byte Ed25519 pubkey from the
                                merchant's `.well-known/icp` discovery.
    :param tolerance_seconds: Replay window (default 300s).
    :param now_seconds: Override "now" for testing.
    :returns: The verified envelope dict.
    """
    if now_seconds is None:
        now_seconds = int(time.time())

    # 1. Timestamp window.
    ts_header = _normalized_header_lookup(headers, "x-icp-timestamp")
    if ts_header is None:
        raise ICPError("channel.signature_invalid", "missing X-ICP-Timestamp header")
    try:
        ts = int(ts_header)
    except (TypeError, ValueError) as e:
        raise ICPError(
            "channel.signature_invalid",
            f"invalid X-ICP-Timestamp: {ts_header}",
        ) from e
    if abs(now_seconds - ts) > tolerance_seconds:
        raise ICPError(
            "channel.replay",
            f"timestamp {ts} outside ±{tolerance_seconds}s of {now_seconds}",
        )

    # 2. HTTP-layer signature.
    sig_header = _normalized_header_lookup(headers, "x-icp-signature")
    if sig_header is None:
        raise ICPError("channel.signature_invalid", "missing X-ICP-Signature header")
    m = _HTTP_SIG_PATTERN.match(sig_header)
    if not m:
        raise ICPError(
            "channel.signature_invalid",
            "X-ICP-Signature must be ed25519=<hex>",
        )
    http_sig_hex = m.group(1)
    http_material = f"{ts_header}.{method}.{path}.{body}"
    if not verify_ed25519(http_material, http_sig_hex, merchant_pubkey_raw):
        raise ICPError(
            "channel.signature_invalid",
            "HTTP-layer signature verification failed",
        )

    # 3. Body shape.
    try:
        parsed = json.loads(body)
    except (json.JSONDecodeError, TypeError) as e:
        raise ICPError(
            "channel.signature_invalid",
            f"body is not JSON: {e}",
        ) from e
    envelope = parsed.get("envelope") if isinstance(parsed, dict) else None
    signature = parsed.get("signature") if isinstance(parsed, dict) else None
    if not isinstance(envelope, dict) or not isinstance(signature, dict):
        raise ICPError(
            "channel.signature_invalid",
            "body missing {envelope, signature} object pair",
        )
    sig = signature.get("sig")
    if not isinstance(sig, str):
        raise ICPError(
            "channel.signature_invalid",
            "body.signature.sig must be a hex string",
        )

    # 4. Envelope signature over canonical bytes.
    envelope_canonical = canonical_json(envelope)
    if not verify_ed25519(envelope_canonical, sig, merchant_pubkey_raw):
        raise ICPError(
            "channel.signature_invalid",
            "envelope signature verification failed",
        )

    return envelope

File: crates/stateset-icp-client/src/webhook.rs
//! ICPIP-0005 receiver helpers — `verify_webhook` is the Stripe-style
//! one-call validator for inbound webhook events.
//!
//! Hand it the raw HTTP body bytes, request headers, HTTP method/path,
//! and the merchant's raw 32-byte Ed25519 pubkey from `.well-known/icp`.
//! Get back the parsed `EventEnvelope` JSON value, OR an
//! [`Error::Icp`] with a `channel.*` code that maps directly to HTTP
//! status.
//!
//! Performs every check ICPIP-0005 §6 mandates:
//!   1. HTTP timestamp within ±`tolerance_seconds` (default 300) →
//!      `channel.replay` on miss.
//!   2. HTTP-layer `X-ICP-Signature: ed25519=<hex>` verifies
//!      cryptographically against `<ts>.<method>.<path>.<body>`.
//!   3. Body parses as `{envelope, signature}`.
//!   4. Envelope signature verifies against the merchant pubkey over
//!      the envelope's canonical JSON bytes.
//!
//! Mirrors the `JavaScript` and `Python` SDK helpers byte-for-byte.

use crate::Error;
use crate::canonical::canonical_json;
use crate::identity::verify_ed25519;
use serde_json::Value;
use std::time::{SystemTime, UNIX_EPOCH};

/// Options controlling [`verify_webhook`] behavior.
#[derive(Debug, Clone, Copy)]
pub struct VerifyWebhookOptions {
    /// Allowed clock skew between sender and receiver, in seconds.
    /// Defaults to 300 (ICPIP-0005 §6 mandate).
    pub tolerance_seconds: u64,
    /// Override "now" for testing. `None` uses `SystemTime::now()`.
    pub now_seconds: Option<u64>,
}

impl Default for VerifyWebhookOptions {
    fn default() -> Self {
        Self { tolerance_seconds: 300, now_seconds: None }
    }
}

/// Verify an inbound webhook and return its parsed `EventEnvelope`.
///
/// `headers` is any iterable of `(name, value)` pairs (typically your
/// HTTP library's header collection). Lookup is case-insensitive.
///
/// Returns `Err(Error::Icp { code: "channel.*", … })` on any failure.
pub fn verify_webhook<H>(
    body: &str,
    headers: H,
    method: &str,
    path: &str,
    merchant_pubkey_hex: &str,
    opts: VerifyWebhookOptions,
) -> Result<Value, Error>
where
    H: IntoIterator,
    H::Item: HeaderPair,
{
    // Collect headers once with normalized lowercase keys.
    let mut header_map: Vec<(String, String)> = Vec::with_capacity(8);
    for pair in headers {
        let (k, v) = pair.into_pair();
        header_map.push((k.to_ascii_lowercase(), v));
    }
    let lookup = |name: &str| -> Option<&str> {
        let lower = name.to_ascii_lowercase();
        header_map.iter().find_map(|(k, v)| (*k == lower).then_some(v.as_str()))
    };

    // 1. Timestamp window.
    let ts_header = lookup("x-icp-timestamp").ok_or_else(|| Error::Icp {
        code: "channel.signature_invalid".to_string(),
        message: "missing X-ICP-Timestamp header".to_string(),
    })?;
    let ts: u64 = ts_header.parse().map_err(|_| Error::Icp {
        code: "channel.signature_invalid".to_string(),
        message: format!("invalid X-ICP-Timestamp: {ts_header}"),
    })?;
    let now = opts
        .now_seconds
        .unwrap_or_else(|| SystemTime::now().duration_since(UNIX_EPOCH).map_or(0, |d| d.as_secs()));
    let drift = now.abs_diff(ts);
    if drift > opts.tolerance_seconds {
        return Err(Error::Icp {
            code: "channel.replay".to_string(),
            message: format!("timestamp {ts} outside ±{}s of {now}", opts.tolerance_seconds,),
        });
    }

    // 2. HTTP-layer signature.
    let sig_header = lookup("x-icp-signature").ok_or_else(|| Error::Icp {
        code: "channel.signature_invalid".to_string(),
        message: "missing X-ICP-Signature header".to_string(),
    })?;
    let http_sig_hex = sig_header.strip_prefix("ed25519=").ok_or_else(|| Error::Icp {
        code: "channel.signature_invalid".to_string(),
        message: "X-ICP-Signature must be ed25519=<hex>".to_string(),
    })?;
    let http_material = format!("{ts_header}.{method}.{path}.{body}");
    if !verify_ed25519(http_material.as_bytes(), http_sig_hex, merchant_pubkey_hex) {
        return Err(Error::Icp {
            code: "channel.signature_invalid".to_string(),
            message: "HTTP-layer signature verification failed".to_string(),
        });
    }

    // 3. Body shape.
    let parsed: Value = serde_json::from_str(body).map_err(|e| Error::Icp {
        code: "channel.signature_invalid".to_string(),
        message: format!("body is not JSON: {e}"),
    })?;
    let envelope = parsed.get("envelope").cloned().ok_or_else(|| Error::Icp {
        code: "channel.signature_invalid".to_string(),
        message: "body missing 'envelope'".to_string(),
    })?;
    let envelope_sig = parsed
        .get("signature")
        .and_then(|s| s.get("sig"))
        .and_then(Value::as_str)
        .ok_or_else(|| Error::Icp {
            code: "channel.signature_invalid".to_string(),
            message: "body missing signature.sig".to_string(),
        })?
        .to_string();

    // 4. Envelope signature over canonical bytes.
    let envelope_canonical = canonical_json(&envelope)?;
    if !verify_ed25519(envelope_canonical.as_bytes(), &envelope_sig, merchant_pubkey_hex) {
        return Err(Error::Icp {
            code: "channel.signature_invalid".to_string(),
            message: "envelope signature verification failed".to_string(),
        });
    }

    Ok(envelope)
}

/// Trait implemented by everything that can be coerced into a header
/// (name, value) string pair. Lets `verify_webhook` accept `Vec<(String,
/// String)>`, `HashMap<String, String>`, `&[(&str, &str)]`, and the
/// header types of popular HTTP crates without taking a dependency on
/// any of them.
pub trait HeaderPair {
    /// Convert into an owned `(name, value)` pair.
    fn into_pair(self) -> (String, String);
}

impl HeaderPair for (String, String) {
    fn into_pair(self) -> (String, String) {
        self
    }
}

impl HeaderPair for (&str, &str) {
    fn into_pair(self) -> (String, String) {
        (self.0.to_string(), self.1.to_string())
    }
}

impl HeaderPair for (&String, &String) {
    fn into_pair(self) -> (String, String) {
        (self.0.clone(), self.1.clone())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Identity;
    use serde_json::json;

    // Fixed merchant keypair so tests are deterministic.
    const MERCHANT_SEED: &str = "5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a";

    fn merchant() -> Identity {
        Identity::from_seeds_hex(MERCHANT_SEED, MERCHANT_SEED).unwrap()
    }

    fn merchant_pubkey_hex() -> String {
        hex::encode(merchant().ed_pubkey())
    }

    fn sample_envelope() -> Value {
        json!({
            "v": "icp-1.0",
            "event_id": "icp_evt_test001",
            "event_type": "settlement.released",
            "channel_id": "icp_ch_test001",
            "sequence": 1,
            "originated_at": "2026-05-12T15:22:09.000Z",
            "source": "aid:v1:zMerchantTest",
            "target": "aid:v1:zAgentTest",
            "payload": {
                "settlement_id": "icp_set_abc",
                "escrow_id": "0xabc",
                "amount": {"amount": "29.99", "currency": "USDC"},
                "final_state": "released"
            },
            "previous_event_id": null,
            "delivery_attempt": 1
        })
    }

    fn forge(
        envelope: Value,
        now: u64,
        method: &str,
        path: &str,
    ) -> (String, Vec<(String, String)>) {
        let m = merchant();
        let env_canonical = canonical_json(&envelope).unwrap();
        let env_sig = m.sign_hex(env_canonical.as_bytes());
        let body = format!(
            "{{\"envelope\":{env_canonical},\"signature\":{{\"alg\":\"ed25519\",\"kid\":\"{}\",\"sig\":\"{env_sig}\"}}}}",
            envelope["source"].as_str().unwrap(),
        );
        let ts = now.to_string();
        let material = format!("{ts}.{method}.{path}.{body}");
        let http_sig = m.sign_hex(material.as_bytes());
        let headers = vec![
            ("content-type".to_string(), "application/json".to_string()),
            ("x-icp-timestamp".to_string(), ts),
            ("x-icp-signature".to_string(), format!("ed25519={http_sig}")),
            ("x-icp-channel-id".to_string(), envelope["channel_id"].as_str().unwrap().to_string()),
        ];
        (body, headers)
    }

    fn opts_at(now: u64) -> VerifyWebhookOptions {
        VerifyWebhookOptions { now_seconds: Some(now), tolerance_seconds: 300 }
    }

    #[test]
    fn happy_path_returns_parsed_envelope() {
        let now = 1_900_000_000;
        let (body, headers) = forge(sample_envelope(), now, "POST", "/icp/events");
        let env = verify_webhook(
            &body,
            headers,
            "POST",
            "/icp/events",
            &merchant_pubkey_hex(),
            opts_at(now),
        )
        .expect("should verify");
        assert_eq!(env["event_id"], "icp_evt_test001");
        assert_eq!(env["payload"]["final_state"], "released");
    }

    #[test]
    fn tampered_body_rejected() {
        let now = 1_900_000_000;
        let (body, headers) = forge(sample_envelope(), now, "POST", "/icp/events");
        let tampered = body.replace("29.99", "99.99");
        let err = verify_webhook(
            &tampered,
            headers,
            "POST",
            "/icp/events",
            &merchant_pubkey_hex(),
            opts_at(now),
        )
        .unwrap_err();
        match err {
            Error::Icp { code, .. } => assert_eq!(code, "channel.signature_invalid"),
            other => panic!("unexpected: {other:?}"),
        }
    }

    #[test]
    fn stale_timestamp_rejected_with_channel_replay() {
        let stale = 1_900_000_000;
        let (body, headers) = forge(sample_envelope(), stale, "POST", "/icp/events");
        let now = stale + 600; // outside ±300s
        let err = verify_webhook(
            &body,
            headers,
            "POST",
            "/icp/events",
            &merchant_pubkey_hex(),
            opts_at(now),
        )
        .unwrap_err();
        match err {
            Error::Icp { code, .. } => assert_eq!(code, "channel.replay"),
            other => panic!("unexpected: {other:?}"),
        }
    }

    #[test]
    fn missing_timestamp_rejected() {
        let now = 1_900_000_000;
        let (body, headers) = forge(sample_envelope(), now, "POST", "/icp/events");
        let without_ts: Vec<_> =
            headers.into_iter().filter(|(k, _)| k != "x-icp-timestamp").collect();
        let err = verify_webhook(
            &body,
            without_ts,
            "POST",
            "/icp/events",
            &merchant_pubkey_hex(),
            opts_at(now),
        )
        .unwrap_err();
        match err {
            Error::Icp { code, message } => {
                assert_eq!(code, "channel.signature_invalid");
                assert!(message.contains("X-ICP-Timestamp"));
            }
            other => panic!("unexpected: {other:?}"),
        }
    }

    #[test]
    fn missing_signature_header_rejected() {
        let now = 1_900_000_000;
        let (body, headers) = forge(sample_envelope(), now, "POST", "/icp/events");
        let without_sig: Vec<_> =
            headers.into_iter().filter(|(k, _)| k != "x-icp-signature").collect();
        let err = verify_webhook(
            &body,
            without_sig,
            "POST",
            "/icp/events",
            &merchant_pubkey_hex(),
            opts_at(now),
        )
        .unwrap_err();
        match err {
            Error::Icp { code, .. } => assert_eq!(code, "channel.signature_invalid"),
            other => panic!("unexpected: {other:?}"),
        }
    }

    #[test]
    fn malformed_signature_algorithm_rejected() {
        let now = 1_900_000_000;
        let (body, mut headers) = forge(sample_envelope(), now, "POST", "/icp/events");
        for (k, v) in &mut headers {
            if k == "x-icp-signature" {
                *v = "hmac-sha256=deadbeef".to_string();
            }
        }
        let err = verify_webhook(
            &body,
            headers,
            "POST",
            "/icp/events",
            &merchant_pubkey_hex(),
            opts_at(now),
        )
        .unwrap_err();
        match err {
            Error::Icp { code, .. } => assert_eq!(code, "channel.signature_invalid"),
            other => panic!("unexpected: {other:?}"),
        }
    }

    #[test]
    fn wrong_pubkey_rejected() {
        let now = 1_900_000_000;
        let (body, headers) = forge(sample_envelope(), now, "POST", "/icp/events");
        let other_pubkey = "0102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f20";
        let err = verify_webhook(&body, headers, "POST", "/icp/events", other_pubkey, opts_at(now))
            .unwrap_err();
        match err {
            Error::Icp { code, .. } => assert_eq!(code, "channel.signature_invalid"),
            other => panic!("unexpected: {other:?}"),
        }
    }

    #[test]
    fn case_insensitive_header_lookup() {
        let now = 1_900_000_000;
        let (body, headers) = forge(sample_envelope(), now, "POST", "/icp/events");
        // Upper-case the names.
        let upper: Vec<_> = headers.into_iter().map(|(k, v)| (k.to_uppercase(), v)).collect();
        let env = verify_webhook(
            &body,
            upper,
            "POST",
            "/icp/events",
            &merchant_pubkey_hex(),
            opts_at(now),
        )
        .expect("should verify with upper-case headers");
        assert_eq!(env["event_id"], "icp_evt_test001");
    }

    #[test]
    fn slice_of_str_pairs_supported() {
        let now = 1_900_000_000;
        let (body, headers) = forge(sample_envelope(), now, "POST", "/icp/events");
        // Borrow as (&str, &str) pairs to exercise the HeaderPair impl.
        let refs: Vec<(&str, &str)> =
            headers.iter().map(|(k, v)| (k.as_str(), v.as_str())).collect();
        let env = verify_webhook(
            &body,
            refs,
            "POST",
            "/icp/events",
            &merchant_pubkey_hex(),
            opts_at(now),
        )
        .expect("should verify");
        assert_eq!(env["event_id"], "icp_evt_test001");
    }
}
